packages/flutter/lib/src/widgets/focus_manager.dart

```diff
@@ -22,16 +22,43 @@ import 'framework.dart';
 /// be logged.
 bool debugFocusChanges = false;
 
-bool _focusDebug(String message, [Iterable<String>? details]) {
-  if (debugFocusChanges) {
-    debugPrint('FOCUS: $message');
-    if (details != null && details.isNotEmpty) {
-      for (final String detail in details) {
-        debugPrint('    $detail');
-      }
+// When using _focusDebug, always call it like so:
+//
+// assert((debugFocusChanges && _focusDebug('Blah $foo')) || true);
+//
+// It needs to be inside the assert in order to be removed in release mode, and
+// it needs the boolean conditional so that the interpolation doesn't happen in
+// debug mode unless the debugFocusChanges value is set. It needs the final ||
+// true because otherwise the assertion will be false when debugFocusChanges is
+// false.
+//
+// It will throw a StateError if you try to call it when debugFocusChanges isn't
+// set, or the app is in release mode.
+//
+// If this gets any uglier, then we should replace the String message argument
+// with an Iterable<Object> and do the stringification inside the function.
+bool _focusDebug(String message, [Iterable<Object>? details]) {
+  if (kReleaseMode) {
+    throw StateError(
+      '_focusDebug was called in Release mode. It should always be wrapped in '
+      'an assert. Always call _focusDebug like so:\n'
+      r"  assert((debugFocusChanges && _focusDebug('Blah $foo')) || true);"
+    );
+  }
+  if (!debugFocusChanges) {
+    throw StateError(
+      "_focusDebug was called when debugFocusChanges wasn't set. "
+      'Always call _focusDebug like so:\n'
+      r"  assert((debugFocusChanges && _focusDebug('Blah $foo')) || true);"
+    );
+  }
+  debugPrint('FOCUS: $message');
+  if (details != null && details.isNotEmpty) {
+    for (final Object detail in details) {
+      debugPrint('    $detail');
     }
   }
-  // Return true so that it can be easily used inside of an assert.
+  // Return true so that it can be used inside of an assert.
```

Comment:
😅

Reply:
Yes, it felt like we were crossing a threshold. :-)